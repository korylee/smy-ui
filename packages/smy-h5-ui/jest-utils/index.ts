import Vue from 'vue'
import { Wrapper } from '@vue/test-utils'
import { IN_BROWSER } from '@smy-h5/shared'

Vue.config.productionTip = false

export const delay = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window | Document, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: Wrapper<Vue> | HTMLElement | Window | Document,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0,
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })
  el.dispatchEvent(event)

  return Vue.nextTick()
}

export async function triggerDrag(el: HTMLElement, relativeX: number, relativeY: number) {
  let x = relativeX
  let y = relativeY
  let startX = 0
  let startY = 0
  if (relativeX < 0) {
    startX = Math.abs(relativeX)
    x = 0
  }
  if (relativeY < 0) {
    startY = Math.abs(relativeY)
    y = 0
  }
  await trigger(el, 'touchstart', startX, startY)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await delay(300)
  await trigger(el, 'touchend', x, y)
}

export function mockConsole(method: keyof Console, fn: any = () => {}) {
  const originMethod = console[method] as any
  console[method] = fn
  return () => {
    console[method] = originMethod
  }
}

export function mockGetBoundingClientRect(rect: Partial<DOMRect>): () => void {
  if (IN_BROWSER) {
    const spy = jest.spyOn(Element.prototype, 'getBoundingClientRect')
    spy.mockReturnValue(rect as DOMRect)
    return () => spy.mockRestore()
  }
  return () => {}
}
