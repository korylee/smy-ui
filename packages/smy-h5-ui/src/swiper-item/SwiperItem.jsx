import { createChildrenMixin } from '../_mixins/relation'
import { createNamespace } from '../_utils/vue/create'

import './swiperItem.less'

const [name, bem] = createNamespace('swiper-item')

export default {
  name,
  mixins: [createChildrenMixin('swiper')],
  data: () => ({ offset: 0 }),
  computed: {
    style({ swiper, offset }) {
      const style = {}
      const isVertical = swiper?.vertical
      const size = swiper?.size
      if (size) {
        style[isVertical ? 'height' : 'width'] = `${size}px`
      }
      if (offset) {
        style.transform = `translate${isVertical ? 'Y' : 'X'}(${offset}px)`
      }
      return style
    },
  },
  methods: {
    setOffset(offset) {
      this.offset = offset
    },
  },
  render() {
    const vm = this
    const _c = vm.$createElement

    const { style } = vm
    return _c(
      'div',
      {
        on: vm.$listeners,
        class: bem(),
        style,
      },
      vm._t('default'),
    )
  },
}
