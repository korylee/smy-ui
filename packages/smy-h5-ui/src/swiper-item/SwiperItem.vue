<template>
  <div v-on="$listeners" class="smy-swiper-item" :style="style">
    <slot />
  </div>
</template>

<script>
import { createChildrenMixin } from '../_utils/mixins/relation'

export default {
  name: 'SmySwiperItem',
  mixins: [createChildrenMixin('swiper')],
  data: () => ({
    offset: 0,
  }),
  computed: {
    style() {
      const style = {}
      const isVertical = this.swiper?.vertical
      const size = this.swiper?.size
      if (size) {
        style[isVertical ? 'height' : 'width'] = `${size}px`
      }
      if (this.offset) {
        style.transform = `translate${isVertical ? 'Y' : 'X'}(${this.offset}px)`
      }
      return style
    },
  },
  methods: {
    setOffset(offset) {
      this.offset = offset
    },
  },
}
</script>
