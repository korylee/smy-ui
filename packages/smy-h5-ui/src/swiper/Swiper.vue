<template>
  <div ref="root" :class="bem({ vertical })">
    <div
      ref="track"
      :class="bem('track')"
      :style="style"
      @touchstart.passive="onTouchStart"
      @touchmove="onTouchMove"
      @touchend="onTouchEnd"
      @touchcancel="onTouchEnd"
    >
      <slot />
    </div>
    <slot name="indicator" :index="activeIndex" :length="count">
      <div v-if="indicator" :class="bem('indicator')">
        <i
          v-for="(item, index) of children"
          :key="item._uid"
          :class="bem('indicator-item', { active: activeIndex === index })"
          @click="to(index)"
        />
      </div>
    </slot>
    <slot name="extra" />
  </div>
</template>

<script>
import { createParentMixin } from '../_mixins/relation'
import { toNumber, range, throttle } from '../_utils/shared'
import { toPxNum, doubleRaf, preventDefault, isHidden } from '../_utils/dom'
import { useTouch } from '../_utils/composable/useTouch'
import { props } from './props'
import { createNamespace } from '../_utils/vue/create'
import { useWindowSize } from '../_utils/composable/useWindowSize'
import { PopupMixin } from '../popup/provide'
import { usePageHidden } from '../_utils/composable/usePageHidden'

const [name, bem] = createNamespace('swiper')

export default {
  name,
  mixins: [createParentMixin('swiper'), PopupMixin],
  props,
  data: () => ({
    touch: useTouch(),
    active: 0,
    rect: null,
    moving: false,
    offset: 0,
    touchTime: 0,
    autoplayTimer: null,
    style: {},
  }),
  computed: {
    isCorrectDirection(vm) {
      return vm.touch?.isVertical() === !!vm.vertical
    },
    size({ vertical, height, width, rect }) {
      if (vertical) return toPxNum(height ?? rect?.height ?? 0)
      return toPxNum(width ?? rect?.width ?? 0)
    },
    count({ children }) {
      return children.length
    },
    trackSize({ count, size }) {
      return count * size
    },
    minOffset({ rect, vertical, trackSize }) {
      if (rect) {
        const base = rect[vertical ? 'height' : 'width']
        return base - trackSize
      }
      return 0
    },
    activeIndex({ count, active }) {
      return (active + count) % count
    },
    delta({ touch, vertical }) {
      if (!touch) return 0
      return touch.state[vertical ? 'deltaY' : 'deltaX']
    },
  },
  created() {
    const vm = this

    const resize = throttle(() => vm.init(), 300)
    const windowSize = useWindowSize()
    vm.$watch(() => {
      const { initialIndex, count } = vm
      return [initialIndex, count, windowSize.width]
    }, resize)

    const { stopAutoplay, startAutoplay } = vm
    const pageHidden = usePageHidden()
    vm.$watch(
      () => pageHidden.value,
      (val) => (val ? stopAutoplay() : startAutoplay()),
    )
    vm.$watch('autoplay', (val) => {
      val > 0 ? startAutoplay() : stopAutoplay()
    })
    vm.$once('hook:beforeDestroy', stopAutoplay)
    vm.$on('hook:deactivated', stopAutoplay)
    vm.popupProvider?.$on('show', resize)
    vm.popupProvider?.$on('close', stopAutoplay)
  },
  methods: {
    bem,
    onTouchStart(e) {
      if (!this.touchable || this.children.length <= 1 || e.touches.length > 1) return
      this.touch.start(e)
      this.touchTime = Date.now()
      this.stopAutoplay()
      this.resetPosition()
    },
    onTouchMove(e) {
      const vm = this
      if (!vm.touchable || !vm.moving) return
      vm.touch.move(e)
      if (!vm.isCorrectDirection) return
      const { delta, loop, active, count } = vm

      const isEdgeTouch = !loop && ((active <= 0 && delta > 0) || (active >= count - 1 && delta < 0))
      if (isEdgeTouch) return
      if (vm.isPreventDefault) preventDefault(e)
      if (vm.isStopPropagation) e.stopPropagation()

      vm.move({ offset: delta })
    },
    onTouchEnd() {
      const { delta, size, touchable, moving, touchTime, touch, isCorrectDirection, vertical } = this
      if (!touchable || !moving) return
      const speed = delta / (Date.now() - touchTime)
      const shouldMove = Math.abs(speed) > 0.3 || Math.abs(delta) > +(size / 2).toFixed(2)
      if (shouldMove && isCorrectDirection) {
        let pace = 0
        const offset = touch.state[vertical ? 'offsetY' : 'offsetX']
        if (this.loop) {
          pace = offset > 0 ? (delta > 0 ? -1 : 1) : 0
        } else {
          pace = -Math[delta > 0 ? 'ceil' : 'floor'](delta / size)
        }
        this.move({ pace, isEmit: true })
      } else if (delta) {
        this.move({ pace: 0 })
      }
      this.moving = false
      this.getStyle()
      this.startAutoplay()
    },
    getActive(pace) {
      const { active, count } = this
      if (pace) {
        if (this.loop) {
          return range(active + pace, -1, count)
        }
        return range(active + pace, 0, count - 1)
      }
      return active
    },
    getOffset(active, offset = 0) {
      const { minOffset, loop, size } = this
      let currentPosition = active * size
      if (!loop) {
        currentPosition = Math.min(currentPosition, -minOffset)
      }
      const targetOffset = offset - currentPosition

      if (!loop) {
        return range(targetOffset, minOffset, 0)
      }
      return targetOffset
    },
    getStyle() {
      let offset = 0
      const { vertical, size, count, rect, duration, moving, center, trackSize } = this
      if (!center) {
        offset = this.offset
      } else {
        const diff = (vertical ? rect.height : rect.width) - size
        offset = this.offset + (this.active === count - 1 ? -diff / 2 : diff / 2)
      }
      const crossName = vertical ? 'width' : 'height'
      const crossAxis = toPxNum(this[crossName] ?? rect[crossName] ?? 0)
      this.style = {
        transitionProperty: moving ? 'none' : undefined,
        transitionDuration: `${moving ? 0 : duration}ms`,
        transform: `translate${vertical ? 'Y' : 'X'}(${offset}px)`,
        [vertical ? 'height' : 'width']: trackSize ? `${trackSize}px` : '',
        [crossName]: crossAxis ? `${crossAxis}px` : '',
      }
    },
    resetPosition() {
      this.moving = true
      const { count, active } = this
      if (active <= -1) {
        this.move({ pace: count })
      } else if (active >= count) {
        this.move({ pace: -count })
      }
    },
    reset(cb) {
      this.resetPosition()
      this.touch.reset()
      return doubleRaf(() => {
        this.moving = false
        return cb && cb()
      })
    },
    prev(step = 1) {
      this.reset(() => {
        this.move({ pace: -step, isEmit: true })
      })
    },
    next(step = 1) {
      this.reset(() => {
        this.move({ pace: step, isEmit: true })
      })
    },
    to(index) {
      this.reset(() => {
        const { count, active } = this
        let targetIndex
        if (this.loop && count === index) {
          targetIndex = active === 0 ? 0 : index
        } else {
          targetIndex = index % count
        }
        this.moving = false
        this.move({ pace: targetIndex - active, isEmit: true })
      })
    },
    startAutoplay() {
      const { autoplay } = this
      if (autoplay <= 0 || this.count <= 1) return
      this.stopAutoplay()
      this.autoplayTimer = setTimeout(() => {
        this.next()
        this.startAutoplay()
      }, toNumber(autoplay))
    },
    stopAutoplay() {
      clearTimeout(this.autoplayTimer)
    },
    init(active = this.initialIndex) {
      const { root } = this.$refs
      if (!root) return
      // eslint-disable-next-line vue/valid-next-tick
      const nextTick = isHidden(root) ? this.$nextTick() : Promise.resolve()
      nextTick.then(() => {
        const { count, children } = this
        if (!isHidden(root)) {
          const rect = {
            width: root.offsetWidth,
            height: root.offsetHeight,
          }
          this.rect = rect
        }
        if (count) {
          active = Math.min(count - 1, Number(active))
          if (active === -1) {
            active = count - 1
          }
        }
        this.active = active
        this.moving = true
        this.offset = this.getOffset(active)
        children.forEach((item) => {
          item.setOffset(0)
        })
        this.getStyle()
        this.startAutoplay()
      })
    },
    move({ offset = 0, pace = 0, isEmit = false }) {
      const { count, active, children, minOffset, trackSize } = this
      if (count <= 1) return
      const targetActive = this.getActive(pace)
      const targetOffset = this.getOffset(targetActive, offset)
      if (this.loop) {
        const firstChild = children[0]
        const lastChild = children[count - 1]
        if (firstChild && targetOffset !== minOffset) {
          const rightBound = targetOffset < minOffset
          firstChild.setOffset(rightBound ? trackSize : 0)
        }
        if (lastChild && targetOffset !== 0) {
          const leftBound = targetOffset > 0
          lastChild.setOffset(leftBound ? -trackSize : 0)
        }
      }
      this.active = targetActive
      this.offset = targetOffset
      if (isEmit && active !== targetActive) {
        this.$emit('change', this.activeIndex)
      }
      this.getStyle()
    },
  },
}
</script>

<style lang="less">
@import './swiper.less';
</style>
