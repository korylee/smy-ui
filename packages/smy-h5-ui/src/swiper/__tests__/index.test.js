import { mount } from '@vue/test-utils'
import Vue from 'vue'
import Swiper from '..'
import { delay } from '../../../jest-utils'
import SwiperItem from '../../swiper-item'

const Wrapper = {
  components: { [Swiper.name]: Swiper, [SwiperItem.name]: SwiperItem },
  props: ['loop', 'autoplay', 'indicator'],
  template: `
<smy-swiper
  ref="swiper"
  style="width: 100px; height: 100px"
  v-bind="$props"
  v-on="$listeners"
>
  <smy-swiper-item>1</smy-swiper-item>
  <smy-swiper-item>2</smy-swiper-item>
  <smy-swiper-item>3</smy-swiper-item>
</smy-swiper>`,
}

test('test swiper & swiper-item use', () => {
  Vue.use(Swiper).use(SwiperItem)
  expect(Vue.component(Swiper.name)).toBeTruthy()
  expect(Vue.component(SwiperItem.name)).toBeTruthy()
})

test('test swiper next & prev & to method', async () => {
  const onChange = jest.fn()
  const wrapper = mount(Wrapper, {
    listeners: { change: onChange },
    propsData: { indicator: true },
  })
  const expectIndicatorIndex = (index) =>
    expect(
      wrapper.findAll('.smy-swiper__indicator-item').wrappers[index].classes('smy-swiper__indicator-item--active')
    ).toBe(true)
  await delay(50)
  const {
    swiper: { to, next, prev },
  } = wrapper.vm.$refs

  to(0)
  await delay(100)
  expect(onChange).toHaveBeenCalledTimes(0)
  expectIndicatorIndex(0)
  expect(wrapper.html()).toMatchSnapshot()

  next()
  await delay(100)
  expect(onChange).toHaveBeenLastCalledWith(1)
  expect(wrapper.html()).toMatchSnapshot()
  expectIndicatorIndex(1)

  prev()
  await delay(100)
  expect(onChange).toHaveBeenLastCalledWith(0)
  expect(wrapper.html()).toMatchSnapshot()
  expectIndicatorIndex(0)

  to(2)
  await delay(100)
  expect(onChange).toHaveBeenLastCalledWith(2)
  expectIndicatorIndex(2)
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})
