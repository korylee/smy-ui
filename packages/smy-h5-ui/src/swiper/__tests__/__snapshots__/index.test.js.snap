// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test swiper next & prev & to method 1`] = `
"<div class="smy-swiper" style="width: 100px; height: 100px;">
  <div class="smy-swiper__inner" style="transition-duration: 500ms; transform: translateX(0px); width: 3072px;">
    <div class="smy-swiper-item" style="width: 1024px;">1</div>
    <div class="smy-swiper-item" style="width: 1024px;">2</div>
    <div class="smy-swiper-item" style="width: 1024px;">3</div>
  </div>
  <div class="smy-swiper__indicator"><i class="smy-swiper__indicator-item smy-swiper__indicator-item--active"></i><i class="smy-swiper__indicator-item"></i><i class="smy-swiper__indicator-item"></i></div>
</div>"
`;

exports[`test swiper next & prev & to method 2`] = `
"<div class="smy-swiper" style="width: 100px; height: 100px;">
  <div class="smy-swiper__inner" style="transition-duration: 500ms; transform: translateX(-1024px); width: 3072px;">
    <div class="smy-swiper-item" style="width: 1024px;">1</div>
    <div class="smy-swiper-item" style="width: 1024px;">2</div>
    <div class="smy-swiper-item" style="width: 1024px;">3</div>
  </div>
  <div class="smy-swiper__indicator"><i class="smy-swiper__indicator-item"></i><i class="smy-swiper__indicator-item smy-swiper__indicator-item--active"></i><i class="smy-swiper__indicator-item"></i></div>
</div>"
`;

exports[`test swiper next & prev & to method 3`] = `
"<div class="smy-swiper" style="width: 100px; height: 100px;">
  <div class="smy-swiper__inner" style="transition-duration: 500ms; transform: translateX(0px); width: 3072px;">
    <div class="smy-swiper-item" style="width: 1024px;">1</div>
    <div class="smy-swiper-item" style="width: 1024px;">2</div>
    <div class="smy-swiper-item" style="width: 1024px;">3</div>
  </div>
  <div class="smy-swiper__indicator"><i class="smy-swiper__indicator-item smy-swiper__indicator-item--active"></i><i class="smy-swiper__indicator-item"></i><i class="smy-swiper__indicator-item"></i></div>
</div>"
`;

exports[`test swiper next & prev & to method 4`] = `
"<div class="smy-swiper" style="width: 100px; height: 100px;">
  <div class="smy-swiper__inner" style="transition-duration: 500ms; transform: translateX(-2048px); width: 3072px;">
    <div class="smy-swiper-item" style="width: 1024px;">1</div>
    <div class="smy-swiper-item" style="width: 1024px;">2</div>
    <div class="smy-swiper-item" style="width: 1024px;">3</div>
  </div>
  <div class="smy-swiper__indicator"><i class="smy-swiper__indicator-item"></i><i class="smy-swiper__indicator-item"></i><i class="smy-swiper__indicator-item smy-swiper__indicator-item--active"></i></div>
</div>"
`;
